Book transfer, book non sales, book stock

esterinac opened the pull request:
### Book transfer
Pemindahan buku dari gudang ke perpustakaan/showroom. List perpustakaan dapat ditambahkan di master data perpustakaan.

**User Flow:**
1. Superadmin/admin pemasaran memilih tambah
2. Superadmin/admin pemasaran memilih data tujuan pemindahan, judul buku dan jumlah buku
3. Superadmin/admin pemasaran dapat menghapus data book_transfer jika terjadi kesalahan saat memilih judul dan jumlah buku
4. Superadmin/admin pemasaran/admin gudang dapat melihat buku yang ingin dikeluarkan dengan melihat PDF bon/list buku 
5. Superadmin/admin gudang memilih staf dan deadline untuk progres penyiapan buku
6. Superadmin/admin gudang memulai progres penyiapan buku
7. Superadmin/admin gudang mengakhiri progres penyiapan buku
8. Superadmin/admin gudang dapat mengedit tanggal penyiapan buku
8. Superadmin/admin pemasaran melakukan finalisasi. Saat finalisasi dilakukan pengurangan stok gudang dan catat (insert) ke transaksi buku
9. Selesai

**Demo video:**
1. Memindahkan buku yang belum ada di perpustakaan FIB  

https://user-images.githubusercontent.com/66486312/118953212-d11c9680-b986-11eb-85f2-a82c5d4ff30d.mp4

2. Memindahkan buku yang sebelumnya sudah ada di perpustakaan FIB

https://user-images.githubusercontent.com/66486312/118953248-d7ab0e00-b986-11eb-9f8f-27418fa1b8c0.mp4

3. Memindahkan buku ke perpustakaan teknik

https://user-images.githubusercontent.com/66486312/118953337-ebef0b00-b986-11eb-8a96-02c8c4681044.mp4

4. Memindahkan buku ke showroom

https://user-images.githubusercontent.com/66486312/118953357-f0b3bf00-b986-11eb-9dc9-eba1de165094.mp4

5. Pemindahan tercatat di transaksi buku

![image](https://user-images.githubusercontent.com/66486312/118953487-0fb25100-b987-11eb-8afe-18f95988a370.png)

6. Edit tanggal di pemindahan buku

https://user-images.githubusercontent.com/66486312/118958730-e34d0380-b98b-11eb-816e-ac4ef6dca34a.mp4

7. Hapus book_transfer (saat hapus, data book_transfer di book_transfer_list, book_transfer_user dan book_transaction juga terhapus)

https://user-images.githubusercontent.com/66486312/118959897-ef859080-b98c-11eb-85d0-eff1c1285c5b.mp4

**Contoh data:**
- Tabel book_transfer
![image](https://user-images.githubusercontent.com/66486312/118960023-10e67c80-b98d-11eb-9898-950927eb56c7.png)
- Tabel book_transfer_list (untuk menyimpan buku-buku pemindahan)
![image](https://user-images.githubusercontent.com/66486312/118954870-62d8d380-b988-11eb-943a-a69d3fe02cf5.png)
- Tabel book_transfer_user (untuk menyimpan nama staf gudang yang bertugas)
![image](https://user-images.githubusercontent.com/66486312/118954915-6ec49580-b988-11eb-9f7d-7e497635c74d.png)
- Tabel library_stock_detail (untuk menyimpan stok per perpustakaan)
![image](https://user-images.githubusercontent.com/66486312/118956884-34f48e80-b98a-11eb-96a4-332babdc5c62.png)

### Book non sales
Pengeluaran buku dari gudang yang bersifat gratis/tidak dijual.

**User flow:**
1. Superadmin/admin pemasaran memilih tambah
2. Superadmin/admin pemasaran mengisi nama/alamat tujuan/peminta buku, tipe non penjualan, judul dan jumlah buku
3. Superadmin/admin pemasaran dapat menghapus data book_non_sales jika terjadi kesalahan saat memilih judul dan jumlah buku
4. Superadmin/admin gudang dapat melihat buku yang disiapkan melalui PDF bon/list buku dan menyelesaikan permintaan dengan memilih button centang. Saat menyelesaikan non penjualan dilakukan pengurangan stok gudang dan pencatatan transaksi buku
5. Selesai

**Demo video:**
1. Buku keluar untuk non penjualan tipe presentgift

https://user-images.githubusercontent.com/66486312/118953670-3ec8c280-b987-11eb-9cd1-2ffd949f32f9.mp4

2. Buku keluar untuk non penjualan tipe lain-lain

https://user-images.githubusercontent.com/66486312/118953722-48eac100-b987-11eb-95f6-91b05ada948d.mp4

3. Hapus non penjualan (saat hapus, data book_non_sales di book_non_sales_list dan book_transaction juga terhapus)

https://user-images.githubusercontent.com/66486312/118957712-f01d2780-b98a-11eb-9753-3e3237ff0b8c.mp4

**Contoh data:**
- Tabel book_non_sales
![image](https://user-images.githubusercontent.com/66486312/118955234-b4815e00-b988-11eb-883a-c76168ba98db.png)
- Tabel book_non_sales_list (untuk menyimpan data buku non sales)
![image](https://user-images.githubusercontent.com/66486312/118955273-bd722f80-b988-11eb-9630-832f00a7a92d.png)

### Book stock
- Menambahkan maximal value di revisi supaya stok tidak bisa minus. closed #43 
![image](https://user-images.githubusercontent.com/66486312/118952829-76833a80-b986-11eb-8fc5-41c923ddc6cf.png)

- Membenarkan log revisi stok (+ -)
![image](https://user-images.githubusercontent.com/66486312/118952865-7edb7580-b986-11eb-9b87-5742bdab853e.png)

- Range stok buku gudang 
Sebelumnya filter stok gudang <=50 dan >50. Diubah menjadi range supaya pemilihan angka stok filter menjadi lebih fleksibel (tidak hanya 50).

https://user-images.githubusercontent.com/66486312/118954616-21e0bf00-b988-11eb-9e34-ce4c2d29226a.mp4

- Isi lokasi rak buku di gudang

https://user-images.githubusercontent.com/66486312/118955884-5012ce80-b989-11eb-997a-22e4cdffeaed.mp4

- Chart transaksi buku masuk/keluar
![screencapture-localhost-sigap-book-stock-view-15-2021-05-20-16_06_36](https://user-images.githubusercontent.com/66486312/118952880-8438c000-b986-11eb-9181-dca7d4587486.png)

- Menambahkan retur buku (penukaran buku rusak dari pembeli) dan excel log retur buku
Penambahan retur dilakukan saat ada pembeli buku yang ingin menukar buku rusak dengan buku baru dari gudang (stok gudang berkurang, stok retur/rusak bertambah). Penghapusan retur dilakukan saat ingin menghapus/membuang buku retur/rusak yang sebelumnya ditukar (stok retur berkurang).

**Userflow retur:**
1. Superadmin/admin gudang memilih button retur di index
2. Superadmin/admin gudang memilih penambahan/pengurangan retur, tanggal retur, jumlah retur, dan catatan. 
3. Superadmin/admin gudang memilih button submit. Dilakukan pengurangan stok gudang/retur.

**Demo video retur**
- Tambah retur

https://user-images.githubusercontent.com/66486312/118957867-1642c780-b98b-11eb-81f0-3a6b844b0262.mp4

- Hapus retur

https://user-images.githubusercontent.com/66486312/118955770-33769680-b989-11eb-94ee-624df2cc4ed5.mp4

- Excel stok buku dan retur

https://user-images.githubusercontent.com/66486312/118955711-2659a780-b989-11eb-8f01-981679cf4161.mp4

**Contoh data retur di tabel book_stock_revision:**
![image](https://user-images.githubusercontent.com/66486312/118952975-9adf1700-b986-11eb-8f08-43a56aeb3a52.png)
Penambahan dan penghapusan retur dicatat di tabel book_stock_revision. Retur memiliki value revision_type = “return”. Penambahan retur memiliki type ‘sub’ sedangkan penghapusan retur memiliki type ‘del’.

esterinac committed:
Book transfer

esterinac committed:
Book non sales

esterinac committed:
Fix book stock

- Fix minus stock
- Add book transaction chart
- Add retur
- Fix log revisi

esterinac committed:
Fix book transfer & non sales delete

esterinac commented:
Di-fix di commit c0979f8b65705189cad019735cf3f73075e4fcf3 PR #45 ini ya mas

bagaskarala commented:
mau tanya, library_stock_detail itu dipake dimana ya?

bagaskarala commented:
kalau hapus data book transfer/non_sales mempengaruhi stok buku gak?

esterinac commented on application/modules/book_transfer/controllers/Book_transfer.php:
```diff
@@ -0,0 +1,476 @@
+        // update book stock tapi tabel list buku yg dipindahin blm ada
+        $book_transfer_lists  = $this->book_transfer->fetch_book_transfer_list($book_transfer_id);
+
+        // update stok perpustakaan
+        if($book_transfer->library_id){
+            foreach($book_transfer_lists as $book_transfer_list){
+                $book_stock = $this->book_stock->where('book_id', $book_transfer_list->book_id)->get();
+                $book_stock->library_present += $book_transfer_list->qty;
+                $this->book_stock->where('book_id', $book_transfer_list->book_id)->update($book_stock);
+                // stok detail perpustakaan
+                $library_stock_detail = $this->book_stock->get_one_library_stock($book_stock->book_stock_id,$book_transfer->library_id);
+                // ini bermasalah
+                if($library_stock_detail){
+                    $library_stock_detail->library_stock += $book_transfer_list->qty;
+                    $this->db->set('library_stock', $library_stock_detail->library_stock);
+                    $this->db->where('library_id', $library_stock_detail->library_id);
+                    $this->db->where('book_stock_id', $book_stock->book_stock_id);
+                    $this->db->update('library_stock_detail');
+                }
```
> mau tanya, library_stock_detail itu dipake dimana ya?

library_stock_detail dipake untuk mencatat stok buku di suatu perpustakaan saat ada pemindahan ke perpustakaan mas. Misal pernah ada pemindahan buku **'40 Tahun (1966-2005) Hubungan Indonesia Korea Selatan'** ke **Perpustakaan Fakultas Ilmu Budaya UGM** dan **Perpustakaan Fakultas Teknik UGM**. Nanti di view detail stok buku dapat dilihat jumlah stok di tiap perpustakaan seperti pada gambar di bawah ini. 
![image](https://user-images.githubusercontent.com/66486312/119211918-5dda6800-badf-11eb-834d-82f1c8cfe9b4.png)

Contoh datanya seperti ini mas. Buku **'40 Tahun (1966-2005) Hubungan Indonesia Korea Selatan'** memiliki `book_stock_id`=18.
Tabel library_stock_detail:
![image](https://user-images.githubusercontent.com/66486312/119212020-0d173f00-bae0-11eb-9340-a461143e35e7.png)
Tabel library:
![image](https://user-images.githubusercontent.com/66486312/119212021-14d6e380-bae0-11eb-90c8-a00c3de7ee88.png)

esterinac replied:
btw kan ada komen `// ini bermasalah` di atas line `if($library_stock_detail)` ya mas, itu sudah bisa kok mas kemarin lupa hapus komennya. di demo video yang di atas ini juga sudah ngetest dan sudah bisa update stok di berbagai perpustakaan (baik untuk perpustakaan yang belum memiliki buku tertentu mau pun perpustakaan yang sudah memiliki buku tertentu).

bagaskarala replied:
okey nice

esterinac commented:
> kalau hapus data book transfer/non_sales mempengaruhi stok buku gak?

kalau sekarang, hapus data book transfer/non sales tidak mempengaruhi stok buku mas. 

atau kalau dibikin jadi seperti ini boleh tidak mas? book transfer/non sales hanya bisa dihapus saat belum melakukan finalisasi/selesai. jadi kalau book transfer/non sales sudah selesai, datanya tidak bisa dihapus supaya saat menghapus data transfer/non_sales, tidak perlu menghapus data transaksi dan mengubah stok buku juga.

bagaskarala commented:
> > kalau hapus data book transfer/non_sales mempengaruhi stok buku gak?
> 
> kalau sekarang, hapus data book transfer/non sales tidak mempengaruhi stok buku mas. 
> 
> atau kalau dibikin jadi seperti ini boleh tidak mas? book transfer/non sales hanya bisa dihapus saat belum melakukan finalisasi/selesai. jadi kalau book transfer/non sales sudah selesai, datanya tidak bisa dihapus supaya saat menghapus data transfer/non_sales, tidak perlu menghapus data transaksi dan mengubah stok buku juga.

Iya coba gitu,,

Soalnya kalo yg lama2 bisa dihapus, nanti stok/transaksi berpotensi jd ga konsisten.

Coba sekalian testing konsistensi datanya ya..

esterinac committed:
Fix delete non sales

- Hanya bisa hapus saat status != finish
- Tambah akses untuk level admin_pemasaran di sidebar non_sales

esterinac committed:
Fix delete book transfer

- Hanya bisa hapus saat status != finish

esterinac committed:
Fix delete book receive

- Book receive tidak dapat dihapus
- Merapikan tampilan (hapus tulisan aksi di index, tabel responsive)

esterinac committed:
Fix index book stock

Hapus tulisan aksi di view index book stock

esterinac commented:
> > > kalau hapus data book transfer/non_sales mempengaruhi stok buku gak?
> > 
> > 
> > kalau sekarang, hapus data book transfer/non sales tidak mempengaruhi stok buku mas.
> > atau kalau dibikin jadi seperti ini boleh tidak mas? book transfer/non sales hanya bisa dihapus saat belum melakukan finalisasi/selesai. jadi kalau book transfer/non sales sudah selesai, datanya tidak bisa dihapus supaya saat menghapus data transfer/non_sales, tidak perlu menghapus data transaksi dan mengubah stok buku juga.
> 
> Iya coba gitu,,
> 
> Soalnya kalo yg lama2 bisa dihapus, nanti stok/transaksi berpotensi jd ga konsisten.
> 
> Coba sekalian testing konsistensi datanya ya..

Baik mas, sudah dibikin seperti yang ini

esterinac committed:
Delete unused comments

bagaskarala commented:
![image](https://user-images.githubusercontent.com/14112952/119232146-4a1a1a80-bb4e-11eb-8ecc-3d35d10ceef1.png)

ini gak jadi bisa edit buku setelah final?

bagaskarala commented:
di pesanan buku ada error gini, ada yg perlu disesuaikan?
![image](https://user-images.githubusercontent.com/14112952/119233142-bb0f0180-bb51-11eb-9c25-30fb5eee0252.png)

esterinac commented:
> ![image](https://user-images.githubusercontent.com/14112952/119232146-4a1a1a80-bb4e-11eb-8ecc-3d35d10ceef1.png)
> 
> ini gak jadi bisa edit buku setelah final?

oh iya mas harusnya tetep bisa karena yang diubah hanya tanggal untuk penyiapannya saja, jadi harusnya tidak mempengaruhi data di transaksi buku (karena data tanggal di transaksi mencatat waktu selesai).

esterinac committed:
Fix edit pemindahan buku

Bisa edit book transfer kapan pun

esterinac committed:
Fix filter_book_request

Hanya menampilkan faktur gudang di index book_request

esterinac commented:
> di pesanan buku ada error gini, ada yg perlu disesuaikan?
> ![image](https://user-images.githubusercontent.com/14112952/119233142-bb0f0180-bb51-11eb-9c25-30fb5eee0252.png)

oh iya ada error terima kasih mas, sudah di-fix ya mas

bagaskarala reviewed (approved):
okesip LGTM
